ENH: Cythonize peak_prominences

lagru opened the pull request:
* The inner loop of `peak_prominences` (introduced with #8264) that uses a vectorized approach to find a peak's bases is replaced with a **new iterative, cythonized solution**. This yields significant speed gains as demonstrated in this [notebook](http://nbviewer.jupyter.org/urls/gist.githubusercontent.com/lagru/add10aa2ecb62966297ca0d4a7e21b92/raw/b9a7b9a2749729ed977c2f6ac61fb9241597fdb4/eval_peak_prominences.ipynb). The input validation is still performed in Python.

* Furthermore a bug with the old implementation of `peak_prominences` is fixed that in some edge cases (see **new unit tests** and this [notebook](http://nbviewer.jupyter.org/urls/gist.githubusercontent.com/lagru/add10aa2ecb62966297ca0d4a7e21b92/raw/b9a7b9a2749729ed977c2f6ac61fb9241597fdb4/eval_peak_prominences.ipynb)) led the function to choose the wrong candidate as a peak's base.

* Includes a **clearer description of edge cases** (adjacent peaks with the same height) in the Notes section of `peak_prominences`.

lagru commented on scipy/signal/_peak_finding_utils.pyx:
```diff
@@ -74,3 +74,91 @@ def _argmaxima1d(np.float64_t[:] x not None):
+
+    .. versionadded:: 1.1.0
+    """
+    cdef np.ndarray[np.float64_t, ndim=1] prominences
```
It wasn't a style choice but merely me being unaware of this syntax. Thanks for pointing this out, I'll fix it.

perimosocordiae commented on scipy/signal/_peak_finding.py:
```diff
@@ -354,60 +358,9 @@ def peak_prominences(x, peaks, wlen=None):
-        right_bases[i] += wleft
-
-    return prominences, left_bases, right_bases
+    peaks = np.asarray(peaks, dtype=np.intp)
```
You already call `np.asarray(peaks)` on line 343.

lagru replied:
Yeah, but in that case i call it without the `dtype=np.intp` argument. 
In the first case I do it in order to test for peaks size and dimension. If I would add the typecast here the check `np.issubdtype(peaks.dtype, np.integer)` after this would be meaningless.

I think this is a little ugly but I couldn't find a more elegant solution that checks for integerness of peaks and ensures that the correct int-size is used on all architectures.

perimosocordiae replied:
How about
```
peaks = np.asarray(peaks).astype(np.intp, order='C', casting='safe', subok=False, copy=False)
```

on line 343. This would raise an error if `peaks.dtype` isn't safely castable to `intp` (essentially the same as your manual check) and also makes sure it's stored contiguously.

perimosocordiae commented on scipy/signal/_peak_finding_utils.pyx:
```diff
@@ -74,3 +74,91 @@ def _argmaxima1d(np.float64_t[:] x not None):
+    # Allocate returned objects
+    prominences = np.empty(peaks.shape[0], dtype=np.float64)
+    left_bases = np.empty(peaks.shape[0], dtype=np.intp)
+    right_bases = np.empty(peaks.shape[0], dtype=np.intp)
```
Alternately, use `np.empty_like`

lagru replied:
I thought about it but I'm not sure whether

* it would actually be faster
* I wouldn't have problems with the memory layout because by default `empty_like` tries to match the memory layout as close as possible... which I think would have been a problem if `x` was a non-contiguous array.

lagru replied:
Furthermore there is no measurable difference between using `peaks.shape[0]` repeatedly and storing it inside a typed variable. I find the first case more readable.

perimosocordiae replied:
Works for me.

perimosocordiae commented on scipy/signal/_peak_finding.py:
```diff
@@ -337,11 +339,13 @@ def peak_prominences(x, peaks, wlen=None):
     >>> plt.show()
     """
-    x = np.asarray(x)
+    x = np.asarray(x, dtype=np.float64)
```
Might be worth trying `np.ascontiguousarray` here and using the contiguous memoryview syntax in the Cython code (`[::1]`).

lagru replied:
I wasn't sure about the tradeoffs in that case. If I understand the what I read in [Cython's](http://docs.cython.org/en/latest/src/userguide/memoryviews.html#c-and-fortran-contiguous-memoryviews) and NumPy's documentation as well as [here](https://stackoverflow.com/questions/26998223/what-is-the-difference-between-contiguous-and-non-contiguous-arrays#26999092) correctly this means:

* Indexing of contiguous arrays is slightly (?) faster with the option `[::1]`.
* If the input arrays aren't contiguous there is an time penalty because it needs to be converted / copied? How common is this case? I have a feeling that his case would be rare and this optimization might be worth it.

Are there any other implications if i use `np.ascontiguousarray` and `[::1]` I'm not aware of?

perimosocordiae replied:
`ascontiguousarray` is free if:
 * we're going to make a copy anyway (by changing dtype, or coercing a list)
 * we're already contiguous

Note that in the first case we'd have to do the dtype casting before `ascontiguousarray`, to avoid copying twice.

I'd try it out and see what the benchmarks show. The generated C code with `[::1]` memoryviews should also be slightly different.

lagru commented:
> If there are no such benchmarks in our suite, then adding them (or opening an issue to add them), might be helpful to have a common / general perf metric moving forward.

No, currently there are no benchmarks.
I don't have any experience with `asv` and benchmarking in general (beyond what you see in the notebook). But if you guys think this would really be helpful I might look into it and try to come up with a few meaningful benchmarking scenarios after this is done.
Looking at existing benchmarks in [scipy/benchmarks](https://github.com/scipy/scipy/tree/master/benchmarks) and the notebook content it seems like I'm already halfway there...

lagru commented on scipy/signal/_peak_finding_utils.pyx:
```diff
@@ -74,3 +74,92 @@ def _argmaxima1d(np.float64_t[:] x not None):
+    left_bases = np.empty(peaks.shape[0], dtype=np.intp)
+    right_bases = np.empty(peaks.shape[0], dtype=np.intp)
+
+    for peak_nr in range(peaks.shape[0]):
```
Would it be possible to wrap this loop in a `with nogil:` statement? The `raise` statements would have to be moved outside but with a `raise_error_flag = True; break` the same functionality could easily be achieved. Cython's documentation says [it is safe to release the GIL for memoryviews](http://docs.cython.org/en/latest/src/userguide/memoryviews.html#memoryviews-and-the-gil) (see also [here](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#acquiring-and-releasing-the-gil)). However I'm not sure whether it is safe to release the GIL when indexing memoryviews that point to memory of valid Python objects (see input arrays).

larsoner replied:
> I'm not sure whether it is safe to release the GIL when indexing memoryviews that point to memory of valid Python objects

I suspect this would be a very common use case for typed memoryviews, so if the Cython docs don't list a (prominent) caveat about it, I suspect it should be fine. But if you to ping the Cython devs to ask (and perhaps offer to update their docs to make a note of the answer!) that would be the safest route, unless someone here already knows.

lagru replied:
Good idea. I'll look into it. :+1:
For now, I won't add a `with nogil` until I'm clear on this.

lagru replied:
I can find several other places in SciPy's code base that show a comparable usage: e.g. [_kendal_dis](https://github.com/scipy/scipy/blob/master/scipy/stats/_stats.pyx#L84) or [_colloc](https://github.com/scipy/scipy/blob/636c216eaedf8a570e55ee93a8082d306f772816/scipy/interpolate/_bspl.pyx#L389).

@ev-br (I hope its okay to ping you), the last example was committed by you. Am I right in assuming this is safe to do?

perimosocordiae replied:
+1 for using nogil memoryviews that point to numpy array data. I'm not a Cython dev, but I've used that trick a bunch of times with no issues.

lagru commented on scipy/signal/_peak_finding.py:
```diff
@@ -337,77 +346,38 @@ def peak_prominences(x, peaks, wlen=None):
+    if peaks.size == 0:
+        # The astype statement below would raise a TypeError for empty arrays
+        # which default to dtype np.float64 thus handle this case separately
+        peaks = np.array([], dtype=np.intp)
```
I'm not to happy with the verbose input validation this function has right now. E.g. the handling of `peaks` above. Do you guys see a way to simplify this? Is the complexity in this case worth it?

My thoughts where that in Python there is always the option to be very liberal with the input arguments. However I think user-facing code should fail early and with descriptive error messages. Basically fail for everything instead of converting implicitly.

lagru replied:
Question: is there a better way better way to check if `peaks` can be safely converted to `np.intp`? That would make it possible to simply use `np.asarray(peaks, order='C', dtype=np.intp)` if I new that type conversion would be safe.

larsoner replied:
I think this is an okay way to do it. I don't immediately see a better one.

lagru commented:
[Updated docstring and example as HTML](https://6135-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.peak_prominences.html)

perimosocordiae commented on scipy/signal/_peak_finding.py:
```diff
@@ -336,78 +345,62 @@ def peak_prominences(x, peaks, wlen=None):
-        right_bases[i] += wleft
+    if wlen is None:
+        wlen = -1  # Inner function expects int -> None == -1
+    elif not 1 < wlen:
```
this reads a bit awkwardly to me. How about `elif wlen <= 1:` ?

Alternately, re-order to be:
```python
if wlen is None:
    ...
elif wlen > 1:
   ...
else:
   raise ValueError(...)

lagru replied:
Good catch, the readability can definitely be improved here.

perimosocordiae commented on scipy/signal/_peak_finding_utils.pyx:
```diff
@@ -74,3 +74,98 @@ def _argmaxima1d(np.float64_t[:] x not None):
+
+            prominences[peak_nr] = x[peak] - max(left_min, right_min)
+
+    if raise_error is True:
```
I'm not sure if it affects the generated C code, but it's more pythonic to use `if raise_error:`.

lagru replied:
I don't have strong feelings on this one but I was under the impression that this is analogous to checking for `None` with the `is` statement. I prefer this style usually because I know I won't be surprised later on by some implicit interpretation of empty lists, etc. 

And indeed it does affect the generated C code:
`if raise_error:` ->  `__pyx_t_13 = (__pyx_v_raise_error != 0); if (__pyx_t_13) {`
`if raise_error is True` -> `__pyx_t_13 = ((__pyx_v_raise_error == 1) != 0); if (__pyx_t_13) {`
I think your way is the better one.

perimosocordiae replied:
In this case you're guaranteed the type of `raise_error` (due to the `cdef` above), so you don't have to worry about implicit weirdness.

lagru commented:
@larsoner, I've edited the docstring a little bit, rebased  and all checks are green. If nobody has any issues with this I'd consider this as ready to merge.

By the way, how does GitHub handle the "Changes since last review" if previous commits are no longer available because of a rebase & squash. Can you still see the diff of new changes or only the diff of the entire PR? Is squashing in PRs considered bad practice (excluding the final merge)?

larsoner commented:
> Can you still see the diff of new changes or only the diff of the entire PR?

The latter.

> Is squashing in PRs considered bad practice (excluding the final merge)?

It used to be standard practice to squash to a single commit or a handful of commits. However, with GitHub offer maintainers the option to "squash and merge" as a green button it is less necessary.

Personally I still prefer seeing rebase to seeing merge commits, though. For a +259/-150 PR it's not so bad to have to look again in any case.

perimosocordiae commented on scipy/signal/_peak_finding.py:
```diff
@@ -336,78 +345,62 @@ def peak_prominences(x, peaks, wlen=None):
+            # Empty arrays default to np.float64 but are valid input
+            peaks = np.array([], dtype=np.intp)
+        else:
+            raise TypeError("Cannot safely cast `peaks` to dtype('intp')")
```
This control-flow seems a bit complicated. I'd prefer to check for `peaks.size == 0` before the try/except block, then do the safe `astype` cast without catching the `TypeError`.

lagru replied:
The reasoning is that I expect `peaks` being an empty sequence to be very uncommon. Thus I felt always doing that check might be unnecessary. However, I think the readability trumps the (unnecessary) optimization in this case so I'll change it.

perimosocordiae commented on scipy/signal/_peak_finding.py:
```diff
@@ -336,78 +345,62 @@ def peak_prominences(x, peaks, wlen=None):
+    elif 1 < wlen:
+        # Round up to next positive integer; rounding up to next odd integer
+        # happens implicitly inside the inner function
+        wlen = math.ceil(wlen)
```
You likely want `int(math.ceil(wlen))` here.

lagru replied:
Seems like `math.ceil` should always return an integral value, at least that's what the [docs](https://docs.python.org/3.6/library/math.html#math.ceil) say. So I didn't see the need.
```
>>> type(math.ceil(3.4))
int
>>> type(math.ceil(np.array(2.3)))
int
```

larsoner replied:
Not on 2.7

https://docs.python.org/2.7/library/math.html#math.ceil

lagru replied:
Oh, good catch. But that doesn't seem to be a problem because the unit tests pass even if wlen is a float under 2.7. Maybe that's because of [automatic type conversion](http://docs.cython.org/en/latest/src/reference/language_basics.html#automatic-type-conversion) when passing wlen to the Cython function?

Should we rely on this?

larsoner replied:
It's probably better to be explicit about it.

perimosocordiae commented on scipy/signal/_peak_finding_utils.pyx:
```diff
@@ -148,3 +148,98 @@ def _select_by_peak_distance(np.intp_t[:] peaks not None,
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _peak_prominences(np.float64_t[::1] x not None,
```
You can use a `@cython.nonecheck(False)` directive here instead of manually specifying `not None`.

See https://github.com/cython/cython/wiki/enhancements-compilerdirectives

lagru replied:
I think you mean `@cython.nonecheck(True)`. Using `False` would turn the check off which is off by defaut anyway. 

Using that directive is a good idea for #8594 which features this statement many times. But in this case I like the explicit statements more  as only two arguments use them. Do you agree?

perimosocordiae replied:
Ah, I must have mis-read that directive's docs. Current usage looks good.

lagru commented:
@larsoner, do you want to wait a little longer with merging this? Otherwise I could move on to rebase #8594 so that there is enough time left for reviews before 1.1.0.

larsoner commented:
@perimosocordiae okay for you to merge?

larsoner commented:
Merging to move for 1.1.0, @perimosocordiae if you look again and find issues feel free to comment and @lagru or I can open a PR to fix them

larsoner commented:
Thanks @lagru
